File: cc/13RomanToInteger/RomanToInt.cc
/* 
Given a roman numeral, convert it to an integer.
1 - 10:
I, II, III, IV, V, VI, VII, VIII, IX, X
L = 50, C = 100, D = 500, M = 1000
Input is guaranteed to be within the range from 1 to 3999.
*/

#include <iostream>
#include <string>

using namespace std;

int RomanToInt(string s)
{
	int res = 0;
	for (int i = s.size()-1; i >= 0; i--)
	{
		char c = s[i];
		switch (c) {
		case 'I':
			res += (res >= 5 ? -1 : 1);
			break;
		case 'V':
			res += 5;
			break;
		case 'X':
			res += 10*(res >= 50 ? -1 : 1);
			break;
		case 'L':
			res += 50;
			break;
		case 'C':
			res += 100*(res >= 500 ? -1 : 1);
			break;
		case 'D':
			res += 500;
			break;
		case 'M':
			res += 1000;
			break;
		}
	}
	return res;
}

int main()
{
	string roman = "XII";
	cout<<roman<<": "<<RomanToInt(roman)<<endl;
	return 0;
}

File: py/RomanToInt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
Given a roman numeral, convert it to an integer.

1 ~ 10:
I, II, III, IV, V, VI, VII, VIII, IX, X
L = 50, C = 100, D = 500, M = 1000

Input is guaranteed to be within the range from 1 to 3999.
'''
import time

def romanToInt(s):
	roman = {'M': 1000,'D': 500 ,'C': 100,'L': 50,'X': 10,'V': 5,'I': 1}
	z = 0
	for i in range(0, len(s) - 1):
		if s[i] not in roman or s[i+1] not in roman: # check if it is a valid number
			print "not a valid Roman number"
			return 0
		if roman[s[i]] < roman[s[i+1]]:
			z -= roman[s[i]]
		else:
			z += roman[s[i]]
	return z + roman[s[-1]] # last bit always add

''' test with intToRoman    
s = "VIII"
time0 = time.clock()

print romanToInt(s)
time1 = time.clock()
print "roman to int cost: %f s" % (time1-time0)
'''
